# Lineage-informative microhaplotypes for spatio-temporal surveillance of Plasmodium vivax malaria parasites

This the companion repository for the paper [Lineage-informative microhaplotypes for spatio-temporal surveillance of Plasmodium vivax malaria parasites](https://www.medrxiv.org/content/10.1101/2023.03.13.23287179v1).

The main aim is to further details the methods used for the selection of a candidate microhaplotypes panel for _P. vivax_. In addition, it provides a template framework that can be adapted to different needs such as size, diversity criteria, selection process, etc.

Here we provide two interactive notebooks that go step by step through the selection process, toghether with a number of accessory files to speed up future execution. 
1. In the [first notebook](notebooks/1_evaluating_marker_candidates.ipynb), we scan the _P. vivax_ genome in sliding window and calculates a number of summary statistics (in this example: cardinality, heterozigosity, entropy). Each window roughly maps to a potential microhaplotype, assuming it satisfy a number of criteria (e.g. diversity, number of variants, etc.).
1. The [second notebook](notebooks/2_selection_from_microhaplotype_candidates.ipynb) analyses all the windows together and selects a candidate panel. The selection process is a hard mathematical optimisation problem and here we provide two complementary and effective way to perform the task. The codebase is also modular and can be extended to use different optmisation algorithms.

We used data from a subset of high-quality samples part of the open [MalariaGEN Pv4 dataset](https://www.malariagen.net/resource/30), which contains genome variation data on nearly two-thousands worldwide samples of natural _Plasmodium vivax_ infections. Details on this project, the methods used, and all contributing partners can be found in the key publication: [MalariaGEN et al, Wellcome Open Research 2022, 7:136 https://doi.org/10.12688/wellcomeopenres.17795.1](https://doi.org/10.12688/wellcomeopenres.17795.1). The dataset can be accessed in [a number of ways](https://new_malariagen.staging.sanger.ac.uk/resource/30/) and here we used the the [`malariagen_data` Python package](https://github.com/malariagen/malariagen-data-python), which allows to use the data directly from the cloud and without having to first download them locally. The [Pv4 user guide](https://malariagen.github.io/parasite-data/pv4/Data_access.html) provides all the information on how to use the package as well as some examples to get started.

The notebooks can be run from any computer, including via MyBinder or Google Colab, two  free interactive computing service running in the cloud. Note that the first notebook requires to navigate through hundreds of thousands of genetic variation in thousands of samples and, while the [`malariagen_data` Python package] provides and efficient way to access the data directly on the cloud, the process can still take hours (or days!) depending on the available computing infrastructure. To jump-start the selection process described in the second notebook, we have also provided a number of pre-calculated statistics.

The code contained here has been developed by [Roberto Amato](https://github.com/roamato), [Kathryn Murie](https://github.com/kathryn1995), and [Sasha Siegel](https://github.com/svsiegel).