
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Microhaplotype candidate panel selection &#8212; Lineage-informative microhaplotypes for spatio-temporal surveillance of Plasmodium vivax malaria parasites</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=384b581d" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=36754332"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js?v=afe5de03"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'selection';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Discovery of potential microhaplotype markers in Plasmodium vivax" href="windows.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
  
    <p class="title logo__title">Lineage-informative microhaplotypes for spatio-temporal surveillance of Plasmodium vivax malaria parasites</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="index.html">
                    Lineage-informative microhaplotypes for spatio-temporal surveillance of Plasmodium vivax malaria parasites
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="README.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="windows.html">Discovery of potential microhaplotype markers in <em>Plasmodium vivax</em></a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Microhaplotype candidate panel selection</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://mybinder.org/v2/gh/svsiegel/vivax-mhaps/gh-pages?urlpath=lab/tree/docs/selection.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch onBinder"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img src="_static/images/logo_binder.svg">
  </span>
<span class="btn__text-container">Binder</span>
</a>
</li>
      
      
      
      
      <li><a href="https://colab.research.google.com/github/svsiegel/vivax-mhaps/blob/gh-pages/docs/selection.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch onColab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img src="_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
  </ul>
</div>



<a href="https://github.com/svsiegel/vivax-mhaps" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/selection.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Microhaplotype candidate panel selection</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#environment-and-data">Environment and data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-the-data">Loading the data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#load-the-genomic-data">Load the genomic data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-exploration">Data exploration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#candidate-selection">Candidate selection</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utility-functions">Utility functions</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#greedy">Greedy</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evenly-spaced">Evenly spaced</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#worked-example">Worked example</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="microhaplotype-candidate-panel-selection">
<h1>Microhaplotype candidate panel selection<a class="headerlink" href="#microhaplotype-candidate-panel-selection" title="Link to this heading">#</a></h1>
<p>The aim of this notebook is to analyse all the per-window summary statistics generated in the first notebook, explores panel optimisation and use them to assemble a candidate panel set of markers. The selection process is a challenging mathematical optimisation problem, so here we provide two complementary and effective ways to perform the task. It is worth mentioning that while we show selection methods here, often a subsequent manual curation would be required because of certain constraints, downstream requirements or other considerations (proximity to other markers, reduction of gaps across the genome, low/high diversity regions, or individual assay/panel performance during experimental validation, etc.). The codebase is also modular and can be extended to use different optmisation algorithms if required.</p>
<p>The notebook consists of four parts:</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="#environment-and-data"><span class="std std-ref">Environment and data</span></a></p></li>
<li><p><a class="reference internal" href="#data-exploration"><span class="std std-ref">Data exploration</span></a></p></li>
<li><p><a class="reference internal" href="#candidate-selection"><span class="std std-ref">Candidate selection</span></a> using two different approaches</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="#greedy"><span class="std std-ref">“Greedy”</span></a></p></li>
<li><p><a class="reference internal" href="#evenly-spaced"><span class="std std-ref">Evenly-spaced</span></a></p></li>
</ol>
</li>
<li><p><a class="reference internal" href="#worked-example"><span class="std std-ref">Worked example</span></a></p></li>
</ol>
<p>All the data required to run this notebook, as well as the dataframes/files generated in the <a class="reference internal" href="windows.html"><span class="doc std std-doc">previous notebook</span></a> are also provided pre-calculated in this repository.</p>
<section id="environment-and-data">
<span id="id1"></span><h2>Environment and data<a class="headerlink" href="#environment-and-data" title="Link to this heading">#</a></h2>
<p>This notebook can be run from any computer and can also work from a compute node within Google Cloud, for example via MyBinder or Google Colab, which are free interactive computing services running in a cloud environment. We primarily use common packages that come pre-installed in many computing environments. However, some specialised packages (e.g. <a class="reference external" href="https://github.com/malariagen/malariagen-data-python"><code class="docutils literal notranslate"><span class="pre">malariagen_data</span></code></a>) might require manual installation.</p>
<p>For example, the line below can be uncommented and run on Google Colab to install <code class="docutils literal notranslate"><span class="pre">malariagen-data</span></code> if needed:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="ch">#!pip install -q --no-warn-conflicts malariagen-data==7.14.1</span>
</pre></div>
</div>
</div>
</div>
<p>First, we load all required packages to get started.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load all required packages and print their version</span>
<span class="kn">from</span> <span class="nn">importlib.metadata</span> <span class="kn">import</span> <span class="n">version</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">cycle</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Numpy version: </span><span class="si">{</span><span class="n">version</span><span class="p">(</span><span class="s2">&quot;numpy&quot;</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Pandas version: </span><span class="si">{</span><span class="n">version</span><span class="p">(</span><span class="s2">&quot;pandas&quot;</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Matplotlib version: </span><span class="si">{</span><span class="n">version</span><span class="p">(</span><span class="s2">&quot;matplotlib&quot;</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;scikit-allel version: </span><span class="si">{</span><span class="n">version</span><span class="p">(</span><span class="s2">&quot;scikit-allel&quot;</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;malariagen-data version: </span><span class="si">{</span><span class="n">version</span><span class="p">(</span><span class="s2">&quot;malariagen_data&quot;</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
<span class="kn">import</span> <span class="nn">allel</span>
<span class="kn">from</span> <span class="nn">malariagen_data.pv4</span> <span class="kn">import</span> <span class="n">Pv4</span>

<span class="c1"># Setup plotting backend</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Numpy version: 1.23.5
Pandas version: 2.1.3
Matplotlib version: 3.8.2
scikit-allel version: 1.3.7
malariagen-data version: 7.14.1
</pre></div>
</div>
</div>
</div>
<section id="loading-the-data">
<h3>Loading the data<a class="headerlink" href="#loading-the-data" title="Link to this heading">#</a></h3>
<p>Next, we need to load the per-window summary statistics generated in the previous notebook, and prepare the dataframe <code class="docutils literal notranslate"><span class="pre">windows_df</span></code> for analysis. In this example, we use the pre-calculated statistics file stored in this repo called <code class="docutils literal notranslate"><span class="pre">precomputed/genome_stats.csv</span></code>. This can be changed to use a different dataset if required, simply direct it to your file of choice in the first line below.</p>
<p>We then sort our <code class="docutils literal notranslate"><span class="pre">windows_df</span></code> by chromosome, and then add a column that calculates the midpoint between the <code class="docutils literal notranslate"><span class="pre">window_start</span></code> and <code class="docutils literal notranslate"><span class="pre">window_end</span></code> which will be helpful when we begin to visualise our windows data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">windows_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../precomputed/genome_stats.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">windows_df</span> <span class="o">=</span> <span class="n">windows_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;chrom&#39;</span><span class="p">)</span>
<span class="c1"># Add a supplementary column containing the window midpoint, useful for plotting</span>
<span class="n">windows_df</span> <span class="o">=</span> <span class="n">windows_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">midpoint</span><span class="o">=</span><span class="n">windows_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s2">&quot;window_start&quot;</span><span class="p">,</span> <span class="s2">&quot;window_end&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<span class="n">windows_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>window_start</th>
      <th>window_end</th>
      <th>variant_counts</th>
      <th>unique_allele_counts</th>
      <th>unique_alleles_with_missing_index</th>
      <th>unique_alleles_with_het_index</th>
      <th>unique_allele_frequencies</th>
      <th>unique_allele_count</th>
      <th>entropy</th>
      <th>het</th>
      <th>midpoint</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>PvP01_01_v1</td>
      <td>121301</td>
      <td>121500</td>
      <td>1</td>
      <td>[387   3 225]</td>
      <td>[]</td>
      <td>[1]</td>
      <td>[0.6292682926829268, 0.004878048780487805, 0.3...</td>
      <td>3</td>
      <td>0.685315</td>
      <td>0.470149</td>
      <td>121400.5</td>
    </tr>
    <tr>
      <th>380</th>
      <td>PvP01_01_v1</td>
      <td>634901</td>
      <td>635100</td>
      <td>3</td>
      <td>[  1   6 235   3 252  23   1   1   4  31  51   7]</td>
      <td>[0, 1, 7]</td>
      <td>[3, 6]</td>
      <td>[0.0016260162601626016, 0.00975609756097561, 0...</td>
      <td>12</td>
      <td>1.399306</td>
      <td>0.674974</td>
      <td>635000.5</td>
    </tr>
    <tr>
      <th>381</th>
      <td>PvP01_01_v1</td>
      <td>634951</td>
      <td>635150</td>
      <td>1</td>
      <td>[  8 291   3 313]</td>
      <td>[0]</td>
      <td>[2]</td>
      <td>[0.013008130081300813, 0.47317073170731705, 0....</td>
      <td>4</td>
      <td>0.780273</td>
      <td>0.516893</td>
      <td>635050.5</td>
    </tr>
    <tr>
      <th>382</th>
      <td>PvP01_01_v1</td>
      <td>636251</td>
      <td>636450</td>
      <td>1</td>
      <td>[  3 137   3 472]</td>
      <td>[0]</td>
      <td>[2]</td>
      <td>[0.004878048780487805, 0.22276422764227644, 0....</td>
      <td>4</td>
      <td>0.589552</td>
      <td>0.361303</td>
      <td>636350.5</td>
    </tr>
    <tr>
      <th>383</th>
      <td>PvP01_01_v1</td>
      <td>637951</td>
      <td>638150</td>
      <td>1</td>
      <td>[  1 391   4 219]</td>
      <td>[0]</td>
      <td>[2]</td>
      <td>[0.0016260162601626016, 0.6357723577235772, 0....</td>
      <td>4</td>
      <td>0.698831</td>
      <td>0.468943</td>
      <td>638050.5</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We can also quickly check to see how many unique windows were identified in our previous notebook.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">windows_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>13498
</pre></div>
</div>
</div>
</div>
<p>Using our selection criteria from Notebook 1, we can see that we have <strong>13,498 potential microhaplotype markers</strong> to choose from across the 14 chromosomes.</p>
<p>As this analysis is limited to the core genome, we also need to load the boundaries of those regions, which are provided as part of the MalariaGEN Pv4 data package.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pv_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../supplementary_files/Pv4_regions.bed&quot;</span><span class="p">,</span>
                         <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="c1"># this avoids very long strings</span>
                         <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;chromStart&quot;</span><span class="p">,</span> <span class="s2">&quot;chromEnd&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">])</span>

<span class="c1"># Coordinates are stored as 0-based in the file and need to be adjusted</span>
<span class="n">pv_regions</span><span class="p">[[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">,</span> <span class="s2">&quot;chromEnd&quot;</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
<span class="c1"># Out of convenience, we add a &#39;length&#39; column and filter out all non-core regions</span>
<span class="n">pv_regions</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pv_regions</span><span class="p">[</span><span class="s2">&quot;chromEnd&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">pv_regions</span><span class="p">[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">]</span>
<span class="n">pv_core</span> <span class="o">=</span> <span class="n">pv_regions</span><span class="p">[</span><span class="n">pv_regions</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;Core&quot;</span><span class="p">]</span>
<span class="n">pv_core</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>chromStart</th>
      <th>chromEnd</th>
      <th>name</th>
      <th>length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>PvP01_01_v1</td>
      <td>116542</td>
      <td>677963</td>
      <td>Core</td>
      <td>561421</td>
    </tr>
    <tr>
      <th>3</th>
      <td>PvP01_01_v1</td>
      <td>679790</td>
      <td>903592</td>
      <td>Core</td>
      <td>223802</td>
    </tr>
    <tr>
      <th>6</th>
      <td>PvP01_02_v1</td>
      <td>100156</td>
      <td>162349</td>
      <td>Core</td>
      <td>62193</td>
    </tr>
    <tr>
      <th>8</th>
      <td>PvP01_02_v1</td>
      <td>164088</td>
      <td>745644</td>
      <td>Core</td>
      <td>581556</td>
    </tr>
    <tr>
      <th>11</th>
      <td>PvP01_03_v1</td>
      <td>108062</td>
      <td>630664</td>
      <td>Core</td>
      <td>522602</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Finally, we are loading the identifiers (chromosome and position) of the variant selected in the previous notebook. We will use this at the end to add the details of the variants to each selected window.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variant_coords</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../precomputed/variants_selected.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">variant_coords</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>variant_chrom</th>
      <th>variant_position</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>PvP01_01_v1</td>
      <td>121493</td>
    </tr>
    <tr>
      <th>1</th>
      <td>PvP01_01_v1</td>
      <td>136118</td>
    </tr>
    <tr>
      <th>2</th>
      <td>PvP01_01_v1</td>
      <td>137153</td>
    </tr>
    <tr>
      <th>3</th>
      <td>PvP01_01_v1</td>
      <td>143564</td>
    </tr>
    <tr>
      <th>4</th>
      <td>PvP01_01_v1</td>
      <td>146032</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="load-the-genomic-data">
<h3>Load the genomic data<a class="headerlink" href="#load-the-genomic-data" title="Link to this heading">#</a></h3>
<p>The pre-calculated windows file contain all the information needed for the selection of the panel. This section is only required if you would like to run the last part of this notebook (<a class="reference internal" href="#worked-example"><span class="std std-ref">Worked example</span></a>), where genomic data are required to calculate the allele frequencies of each variant and prepare the input file for <code class="docutils literal notranslate"><span class="pre">paneljudge</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pv4</span> <span class="o">=</span> <span class="n">Pv4</span><span class="p">()</span>
<span class="n">sample_metadata</span> <span class="o">=</span> <span class="n">pv4</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">()</span>
<span class="n">variant_dataset</span> <span class="o">=</span> <span class="n">pv4</span><span class="o">.</span><span class="n">variant_calls</span><span class="p">(</span><span class="n">extended</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pv4_fws</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;../supplementary_files/Pv4_fws.txt&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;t&#39;</span><span class="p">)</span>
<span class="n">sample_metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">sample_metadata</span><span class="p">,</span> <span class="n">pv4_fws</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;outer&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="data-exploration">
<span id="id2"></span><h2>Data exploration<a class="headerlink" href="#data-exploration" title="Link to this heading">#</a></h2>
<p>This section of the notebook is not strictly required for the subsequent panel selection and can be skipped. However, it provides some basic insights into the diversity of the genome that can later be used to make more informed decisions on marker selection.</p>
<p>First, let’s have a look at the distribution of variants per window. <strong>Note the log scale!</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">windows_df</span><span class="p">[</span><span class="s2">&quot;variant_counts&quot;</span><span class="p">],</span><span class="n">bins</span><span class="o">=</span><span class="mi">35</span><span class="p">,</span><span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/9a98824a45a44ba3596b00a3bec3c318c19fa65d4f6fed87cdb41bc1ce0a9549.png" src="_images/9a98824a45a44ba3596b00a3bec3c318c19fa65d4f6fed87cdb41bc1ce0a9549.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">windows_df</span><span class="p">[</span><span class="s2">&quot;variant_counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>variant_counts
1     9605
2     2369
3      729
4      285
5      146
6       95
7       63
8       48
9       32
10      28
11      22
12      15
13      12
14      10
15       2
16       5
17       5
18       3
19       5
20       5
21       2
22       2
24       1
25       1
27       1
28       2
29       1
31       1
32       1
35       1
36       1
Name: count, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">windows_df</span><span class="p">[</span><span class="s2">&quot;variant_counts&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>11974
</pre></div>
</div>
</div>
</div>
<p>Nearly 90% of the windows across the genome only have 1 or 2 variants and thus are not suitable microhaplotype candidates.</p>
<p>Next, let’s have a look at the distribution of heterozygosity across the windows.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">windows_df</span><span class="p">[</span><span class="s2">&quot;het&quot;</span><span class="p">],</span><span class="mi">100</span><span class="p">,</span><span class="n">log</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/974140e3954ba1c3c9de60648fe1e3c3310a1c31e7ff5aa715ff9e420a0c1dc7.png" src="_images/974140e3954ba1c3c9de60648fe1e3c3310a1c31e7ff5aa715ff9e420a0c1dc7.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">windows_df</span><span class="p">[</span><span class="s2">&quot;het&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>3830
</pre></div>
</div>
</div>
</div>
<p>The distribution is largely skewed towards low values, because of the large number of window with only 1 or 2 variants, where the theortical maximum heterozygosity is 0.5. In total, there are 3,830 windows with heterozygosity greater than or equal 0.5.</p>
<p>For a first rough idea of how many candidates we have, we can combine the two criteria. For example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span> <span class="p">(</span><span class="n">windows_df</span><span class="p">[</span><span class="s2">&quot;het&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mf">0.6</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">windows_df</span><span class="p">[</span><span class="s2">&quot;variant_counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">between</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span> <span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1110
</pre></div>
</div>
</div>
</div>
<p>Out of the initial list of 13,498 candidate windows, only 1,110 have between 3 and 10 variants and an heterozygosity above 0.6.</p>
<p>Some of these windows will occour in clusters of high diversity along the genome, so we can then decide to look at the distribution of the 13,498 microhaplotype windows across the core genome using a Manhattan plot. The utility function below will plot each window as a dot and concatenate the core regions across the 14 chromosomes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_man</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">column</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;#276FBF&quot;</span><span class="p">,</span> <span class="s2">&quot;#183059&quot;</span><span class="p">]):</span>

    <span class="n">rcParams</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span>
    <span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;font.size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">12</span>
    <span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;axes.labelsize&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">12</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;xtick.labelsize&quot;</span><span class="p">]</span><span class="o">=</span><span class="mi">12</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;ytick.labelsize&quot;</span><span class="p">]</span><span class="o">=</span><span class="mi">12</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;right&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;top&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;&quot;</span><span class="si">{</span><span class="n">column</span><span class="si">}</span><span class="s1">&quot; per window&#39;</span><span class="p">)</span>

    <span class="c1"># Map chromosome numbers to names</span>
    <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
    <span class="n">chrom_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">,</span> <span class="n">cycle</span><span class="p">(</span><span class="n">palette</span><span class="p">)))</span>

    <span class="c1"># Plot the values</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;midpoint&quot;</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)),</span><span class="n">df</span><span class="p">[</span><span class="n">column</span><span class="p">],</span><span class="n">c</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">chrom_dict</span><span class="p">[</span><span class="n">x</span><span class="p">]),</span> <span class="n">s</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="n">column</span><span class="p">])</span>

    <span class="c1"># Plot the X tick marks</span>
    <span class="n">boundaries</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">shift</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="o">!=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">])</span>
    <span class="n">boundaries</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">boundaries</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">boundaries</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">boundaries</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">))</span>
    <span class="n">ticks</span> <span class="o">=</span> <span class="p">[(</span><span class="n">a</span> <span class="o">+</span> <span class="n">b</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">boundaries</span><span class="p">,</span> <span class="n">boundaries</span><span class="p">[</span><span class="mi">1</span><span class="p">:])]</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">ticks</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">15</span><span class="p">),</span> <span class="n">rotation</span> <span class="o">=</span> <span class="mi">45</span><span class="p">)</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the number of SNPs per microhaplotype window</span>
<span class="n">plot_man</span><span class="p">(</span><span class="n">windows_df</span><span class="p">,</span> <span class="s1">&#39;variant_counts&#39;</span><span class="p">)</span>

<span class="c1"># Rich rainbow colour palette</span>
<span class="c1">#rainbow_palette = [&#39;#C62828&#39;,&#39;#D81B60&#39;,&#39;#8E24AA&#39;,&#39;#5E35B1&#39;,&#39;#3949AB&#39;,&#39;#1E88E5&#39;,&#39;#039BE5&#39;,</span>
<span class="c1">#               &#39;#00ACC1&#39;,&#39;#00897B&#39;,&#39;#43A047&#39;,&#39;#7CB342&#39;,&#39;#C0CA33&#39;,&#39;#FFB300&#39;,&#39;#F4511E&#39;]</span>
<span class="c1">#plot_man(windows_df, &#39;variant_counts&#39;, rainbow_palette) # for a more colorful version</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/9ee682188218bef3805f0116e3528c7809e56ada469979af7877f11b6e674c00.png" src="_images/9ee682188218bef3805f0116e3528c7809e56ada469979af7877f11b6e674c00.png" />
</div>
</div>
<p>Similarly, we can visualise the <strong>global heterozygosity</strong> per microhaplotype window, which indicates the overall level of informativeness contained in a given window. In simple terms, heterozygosity in this context can be described as the likelihood of any two parasites having <strong>different</strong> alleles at each microhaplotype (multiallelic) locus. So the higher the heterozygosity value for any window, the higher the chance that any two global parasite lineages selected from the dataset will be unique from each other. Heterozygosity is a valuable statistic to quantify marker informativeness, especially in the context of being able to capture parasite relatedness relationships from sparse data (such as a genotyping panel).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_man</span><span class="p">(</span><span class="n">windows_df</span><span class="p">,</span> <span class="s1">&#39;het&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/45663a75c944f5303bba41e1e965fd00e6ca2a986c5ee1c0b044e0158136d594.png" src="_images/45663a75c944f5303bba41e1e965fd00e6ca2a986c5ee1c0b044e0158136d594.png" />
</div>
</div>
<p>Another canonical way to quantify “information content” is entropu - for more details see [Entropy (information theory)](<a class="reference external" href="https://en.wikipedia.org/wiki/Entropy_(information_theory)">https://en.wikipedia.org/wiki/Entropy_(information_theory)</a>. We can plot the relationship between heterozygosity and entropy below for comparison.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_het_v_entropy</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>

    <span class="n">rcParams</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span>
    <span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.size&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">12</span>
    <span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;axes.labelsize&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">12</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;xtick.labelsize&#39;</span><span class="p">]</span><span class="o">=</span><span class="mi">10</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;ytick.labelsize&#39;</span><span class="p">]</span><span class="o">=</span><span class="mi">10</span>

    <span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>

    <span class="n">ax1</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;top&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;entropy&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;het&#39;</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Entropy&#39;</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Heterozygosity&#39;</span><span class="p">)</span>

    <span class="n">ax2</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax2</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;top&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">boxplot</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">column</span><span class="o">=</span><span class="s1">&#39;het&#39;</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s1">&#39;variant_counts&#39;</span><span class="p">,</span>
                         <span class="n">rot</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span> <span class="n">grid</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">showfliers</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                         <span class="n">ax</span><span class="o">=</span><span class="n">ax2</span><span class="p">)</span>
    
    <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Number of variants&#39;</span><span class="p">)</span>
    <span class="n">ax2</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">ax2</span><span class="o">.</span><span class="n">title</span><span class="o">.</span><span class="n">set_text</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>

<span class="n">plot_het_v_entropy</span><span class="p">(</span><span class="n">windows_df</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/213eabd36e7dc6fd644d8bac413b88dd12432332efc57af24bda8624aec724a0.png" src="_images/213eabd36e7dc6fd644d8bac413b88dd12432332efc57af24bda8624aec724a0.png" />
</div>
</div>
<p>The left plot indicates that heterozygosity and entropy (both measures of informativeness) are correlated, so for simplicity we have only decided to use heterozygosity as an example here.</p>
<p>The right plot shows that as the number of variants increases, heterozygosity approaches 1 with diminishing returns as variant count increases. This is one of the reasosns why we decided not to select any windows with more than 10 variants, considering that the number of possible haplotypes exponentially increases. The larger the number of possible haplotypes, the more challenging downstream analyses become computationally (as well as the potential implications on error rate that are compounded with each additional variant per window), so it is pragmatic to keep the number of variants per window as low as possible without sacrificing informativeness.</p>
</section>
<section id="candidate-selection">
<span id="id3"></span><h2>Candidate selection<a class="headerlink" href="#candidate-selection" title="Link to this heading">#</a></h2>
<p>Panel selection can be seen as a complex and multidimensional optimisation problem. As shown in <a class="reference external" href="https://doi.org/10.1534/genetics.119.302120">Taylor et al, Genetics 2019</a>, there are several dimensions on which optimal markers should be selected for relatedness estimations. Important considerations from this publication suggest to choose evenly spaced markers across the genome, while simultaneously maximising marker diversity. Additionally, around 100 multiallelic markers were shown to provide sufficient resolution for reconstructing relatedness relationships. From a practical standpoint, there is a trade-off between what is ideal for informativeness maximisation, and even-spacing of markers across the entire genome space. Further, what is reasonable for assay design and implementation of the panel in real-world settings needs to be considered (i.e. number of markers included in an assay, primer design constraints, or how to analyse complex multiallelic data readouts with hundreds/thousands of possible allele combinations). As there are several ways to approach it, here we provide two examples (<strong>greedy</strong> and <strong>evenly spaced</strong>), with the expectation that some level of manual curation will be required to arrive at a final marker set, striking a balance between informativeness, even spacing of markers, and all of the additional considerations needed throughout experimental assay design and marker validation.</p>
<section id="utility-functions">
<h3>Utility functions<a class="headerlink" href="#utility-functions" title="Link to this heading">#</a></h3>
<p>To guarantee that all core regions are represented, we want microhaplotypes to be distributed proportionally to the size of each core region. The utility function below does that and returns the number of expected microhaplotypes per core region, also accounting for rounding errors. This is designed to distribute the number of marker counts proportional to the core region size. Note that this utility function does not space the markers themselves, it simply calculates the number of markers required per core region’s length.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_n_muhap_per_core</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n_muhap_total</span><span class="p">):</span>
    <span class="n">core_length</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>
    <span class="c1"># First try with a naive distribution proportional to the region size</span>
    <span class="n">n_muhap_per_core</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">((</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span><span class="o">/</span><span class="n">core_length</span><span class="p">)</span><span class="o">*</span><span class="n">n_muhap_total</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">)</span>

    <span class="c1"># Check whether the number obtained is what is required (i.e. if there are rounding effects)</span>
    <span class="n">delta</span> <span class="o">=</span> <span class="n">n_muhap_total</span> <span class="o">-</span> <span class="nb">sum</span><span class="p">(</span><span class="n">n_muhap_per_core</span><span class="p">)</span>

    <span class="c1"># Supplement the smallest (or deplete the largest) regions if an adjustment is required</span>
    <span class="k">if</span> <span class="n">delta</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span> <span class="c1"># Less markers than expected, add to the smallest</span>
        <span class="n">smallest</span> <span class="o">=</span> <span class="n">n_muhap_per_core</span><span class="o">.</span><span class="n">nsmallest</span><span class="p">(</span><span class="n">delta</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>
        <span class="n">n_muhap_per_core</span><span class="p">[</span><span class="n">smallest</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">elif</span> <span class="n">delta</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span> <span class="c1"># More markers than expected, remove from the largest</span>
        <span class="n">largest</span> <span class="o">=</span> <span class="n">n_muhap_per_core</span><span class="o">.</span><span class="n">nlargest</span><span class="p">(</span><span class="o">-</span><span class="n">delta</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>
        <span class="n">n_muhap_per_core</span><span class="p">[</span><span class="n">largest</span><span class="p">]</span> <span class="o">-=</span> <span class="mi">1</span>
    
    <span class="k">return</span> <span class="n">n_muhap_per_core</span>
</pre></div>
</div>
</div>
</div>
<p>We can then create a visual representation of marker positions across the genome, where:</p>
<ul class="simple">
<li><p>white = core regions</p></li>
<li><p>red = subtelomeric regions</p></li>
<li><p>orange = internal hypervariable regions</p></li>
<li><p>blue = centromere</p></li>
</ul>
<p>Markers are indicated by vertical lines (black for the selected ones) where the height is proportional to heterozygosity (dotted horizontal line denotes heterozygosity = 0.5).</p>
<p>So we define the palette, and the plotting parameters below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">regions_pal</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Cen&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="s2">&quot;Core&quot;</span><span class="p">:</span><span class="s2">&quot;None&quot;</span><span class="p">,</span> <span class="s2">&quot;In&quot;</span><span class="p">:</span><span class="s2">&quot;orange&quot;</span> <span class="p">,</span> <span class="s2">&quot;Sub&quot;</span><span class="p">:</span><span class="s2">&quot;red&quot;</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_markers</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">markers_list</span><span class="p">,</span> <span class="n">regions</span><span class="p">):</span>

    <span class="n">rcParams</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span>
    <span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.size&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">12</span>
    <span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;axes.labelsize&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">10</span>

    <span class="c1">#plt.rcParams[&#39;xtick.labelsize&#39;]=10</span>
    <span class="c1">#plt.rcParams[&#39;ytick.labelsize&#39;]=10</span>
    
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">sharex</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">sharey</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="n">regions</span><span class="p">[</span><span class="s2">&quot;chromEnd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()])</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_xaxis</span><span class="p">()</span><span class="o">.</span><span class="n">get_major_formatter</span><span class="p">()</span><span class="o">.</span><span class="n">set_scientific</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_xaxis</span><span class="p">()</span><span class="o">.</span><span class="n">set_major_formatter</span><span class="p">(</span><span class="n">mpl</span><span class="o">.</span><span class="n">ticker</span><span class="o">.</span><span class="n">FuncFormatter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">p</span><span class="p">:</span> <span class="nb">format</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="s1">&#39;,&#39;</span><span class="p">)))</span>

    <span class="n">markers</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">markers_list</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">chr_num</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">15</span><span class="p">):</span>
        <span class="n">chr_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;PvP01_</span><span class="si">{</span><span class="n">chr_num</span><span class="si">:</span><span class="s1">02</span><span class="si">}</span><span class="s1">_v1&#39;</span>
    
        <span class="n">chr_regions</span> <span class="o">=</span> <span class="n">regions</span><span class="p">[</span><span class="n">regions</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chr_name</span><span class="p">]</span>
        <span class="n">chr_windows</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chr_name</span><span class="p">]</span>
        <span class="n">chr_markers</span> <span class="o">=</span> <span class="n">markers</span><span class="p">[</span><span class="n">markers</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chr_name</span><span class="p">]</span>

        <span class="n">ax</span><span class="p">[</span><span class="n">chr_num</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="n">chr_num</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;top&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="n">chr_num</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">ax</span><span class="p">[</span><span class="n">chr_num</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">hlines</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">chr_regions</span><span class="p">[</span><span class="s2">&quot;chromEnd&quot;</span><span class="p">]),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#CCCCCC&quot;</span><span class="p">,</span>
                              <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;dotted&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">chr_regions</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">patches</span><span class="o">.</span><span class="n">Rectangle</span><span class="p">((</span><span class="n">r</span><span class="p">[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">],</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">],</span> <span class="mi">2</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;None&#39;</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="n">regions_pal</span><span class="p">[</span><span class="n">r</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]])</span>
            <span class="n">ax</span><span class="p">[</span><span class="n">chr_num</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">add_patch</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>

        <span class="n">ax</span><span class="p">[</span><span class="n">chr_num</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">vlines</span><span class="p">(</span><span class="n">chr_windows</span><span class="p">[</span><span class="s2">&quot;window_start&quot;</span><span class="p">],</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="n">chr_windows</span><span class="p">[</span><span class="s2">&quot;het&quot;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#BBBBBB&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="n">chr_num</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">vlines</span><span class="p">(</span><span class="n">chr_markers</span><span class="p">[</span><span class="s2">&quot;window_start&quot;</span><span class="p">],</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="n">chr_markers</span><span class="p">[</span><span class="s2">&quot;het&quot;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span> 
        <span class="n">ax</span><span class="p">[</span><span class="n">chr_num</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">chr_name</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="s1">&#39;horizontal&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="n">chr_num</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
</pre></div>
</div>
</div>
</div>
<p>Finally, we define an additional function that plots some general statistics (average heterozygosity and inter-marker distance) for a given panel.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_panel_stats</span><span class="p">(</span><span class="n">windows</span><span class="p">):</span>

    <span class="n">rcParams</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span>
    <span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;font.size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">10</span>
    <span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;axes.labelsize&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">12</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;xtick.labelsize&quot;</span><span class="p">]</span><span class="o">=</span><span class="mi">10</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;ytick.labelsize&quot;</span><span class="p">]</span><span class="o">=</span><span class="mi">10</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">windows</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;window_start&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># Guarantee they are sorted by position</span>
    
    <span class="c1"># Left - Heterozygosity per window and average of the candidate panel</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;right&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;top&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Heterozygosity&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

    <span class="n">het_values</span> <span class="o">=</span> <span class="n">windows</span><span class="p">[</span><span class="s2">&quot;het&quot;</span><span class="p">]</span> 
    
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">het_values</span><span class="p">,</span><span class="n">bins</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">het_values</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;darkred&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;dashed&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">het_values</span><span class="o">.</span><span class="n">median</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;navy&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;dashed&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Right - Distance between consecutive markers</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;right&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s2">&quot;top&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance between consecutive windows&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

    <span class="c1"># Calculate distance between widnows on each chromosome separately</span>
    <span class="n">dist_values</span> <span class="o">=</span> <span class="n">windows</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;chrom&quot;</span><span class="p">])[</span><span class="s2">&quot;midpoint&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">diff</span><span class="p">()</span>
    <span class="n">dist_values</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">dist_values</span><span class="p">,</span><span class="n">bins</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">dist_values</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;darkred&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;dashed&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">dist_values</span><span class="o">.</span><span class="n">median</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;navy&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;dashed&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mf">1.5e6</span><span class="p">)</span>

    <span class="c1"># Add text</span>
    <span class="n">_</span><span class="p">,</span> <span class="n">max_ylim</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">()</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">het_values</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">*</span><span class="mf">1.01</span><span class="p">,</span> <span class="n">max_ylim</span><span class="o">*</span><span class="mf">0.97</span><span class="p">,</span> <span class="s2">&quot;Mean: </span><span class="si">{:.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">het_values</span><span class="o">.</span><span class="n">mean</span><span class="p">()),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;darkred&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">het_values</span><span class="o">.</span><span class="n">median</span><span class="p">()</span><span class="o">*</span><span class="mf">1.01</span><span class="p">,</span> <span class="n">max_ylim</span><span class="o">*</span><span class="mf">1.01</span><span class="p">,</span> <span class="s2">&quot;Median: </span><span class="si">{:.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">het_values</span><span class="o">.</span><span class="n">median</span><span class="p">()),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;navy&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">dist_values</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">*</span><span class="mf">1.1</span><span class="p">,</span> <span class="n">max_ylim</span><span class="o">*</span><span class="mf">0.97</span><span class="p">,</span> <span class="s2">&quot;Mean: </span><span class="si">{:,.0f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dist_values</span><span class="o">.</span><span class="n">mean</span><span class="p">()),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;darkred&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">dist_values</span><span class="o">.</span><span class="n">median</span><span class="p">()</span><span class="o">*</span><span class="mf">1.1</span><span class="p">,</span> <span class="n">max_ylim</span><span class="o">*</span><span class="mf">1.01</span><span class="p">,</span> <span class="s2">&quot;Median: </span><span class="si">{:,.0f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dist_values</span><span class="o">.</span><span class="n">median</span><span class="p">()),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;navy&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="greedy">
<span id="id4"></span><h3>Greedy<a class="headerlink" href="#greedy" title="Link to this heading">#</a></h3>
<p>Here we show a heuristic method to maximise informativeness at the expense of other considerations, where within each core region, the proportional number of markers (previously calculated above) are selected only on their window’s heterozygosity statistic, with the highest heterozygosity windows selected regardless of location in the core region. Then, as a second step, you can apply spacing considerations - where the minimum space between markers can be applied (here 1000 bp of space between markers has been set as the default parameter). In the event that the correct number of markers cannot be found for a region, it gives a warning where it fails to retrieve the desired number of markers per core region.</p>
<p>For more information, see <a class="reference external" href="https://en.wikipedia.org/wiki/Greedy_algorithm">Greedy optimisation</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_greedy_muhap_list</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">core</span><span class="p">,</span> <span class="n">n_muhap_per_core</span><span class="p">,</span> <span class="n">min_distance</span><span class="o">=</span><span class="mi">1_000</span><span class="p">):</span>
    
    <span class="n">muhap_list</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Iterate through the core regions one by one</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">row</span><span class="p">),</span> <span class="n">n_core</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">core</span><span class="o">.</span><span class="n">iterrows</span><span class="p">(),</span> <span class="n">n_muhap_per_core</span><span class="p">):</span>
        <span class="c1"># Select all windows in the current region and sort them</span>
        <span class="n">core_windows</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">])</span> <span class="o">&amp;</span>
                           <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;window_start&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">])</span> <span class="o">&amp;</span>
                           <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;window_end&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromEnd&quot;</span><span class="p">])</span> <span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        
        <span class="n">local_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">local_list</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">n_core</span><span class="p">:</span>
            <span class="c1"># Sort by heterozygosity and pick the top - the &quot;greedy&quot; part</span>
            <span class="n">core_windows</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;het&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">core_windows</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;midpoint&quot;</span><span class="p">]):</span>
                <span class="n">local_list</span> <span class="o">+=</span> <span class="p">[</span><span class="n">core_windows</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span> <span class="c1"># Add the top window to the list</span>
                <span class="n">midpoint</span> <span class="o">=</span> <span class="n">core_windows</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;midpoint&quot;</span><span class="p">]</span>
                <span class="c1"># Mask (i.e. ignore in the next iteration) all windows within min_distance from the current one</span>
                <span class="c1"># While the greedy approach doesn&#39;t explicitly consider spacing, this small tweak allows to avoid</span>
                <span class="c1"># considering clusters of variants multiple times</span>
                <span class="n">core_windows</span><span class="o">.</span><span class="n">mask</span><span class="p">(</span> <span class="nb">abs</span><span class="p">(</span><span class="n">core_windows</span><span class="p">[</span><span class="s2">&quot;midpoint&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">midpoint</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_distance</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span> <span class="c1"># If the top marker is NaN, then we are out of options</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Only </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">local_list</span><span class="p">)</span><span class="si">}</span><span class="s1"> out of </span><span class="si">{</span><span class="n">n_core</span><span class="si">}</span><span class="s1"> windows found in </span><span class="se">\</span>
<span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromEnd&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1"> - try decreasing min_distance&#39;</span><span class="p">)</span>
                <span class="k">break</span>
   
        <span class="n">muhap_list</span> <span class="o">+=</span> <span class="n">local_list</span>
    
    <span class="k">return</span> <span class="n">muhap_list</span>
</pre></div>
</div>
</div>
</div>
<p>We can apply the heuristic filters we decided on earlier for choosing only windows that have between 3 and 10 variants per window. <code class="docutils literal notranslate"><span class="pre">windows_subset</span></code> can be changed, depending on needs.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">windows_subset</span> <span class="o">=</span> <span class="n">windows_df</span><span class="p">[</span><span class="n">windows_df</span><span class="p">[</span><span class="s2">&quot;variant_counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">between</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">10</span><span class="p">)]</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">windows_subset</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1426
</pre></div>
</div>
</div>
</div>
<p>We can then create our <strong>greedy</strong> microhaplotype selection list, and ask for a panel with 100 microhaplotypes - this can be changed as needed. You can then specify the minimum distance between markers, here we have set it to 10,000 as an example.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_muhap_total</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">n_muhap_per_core</span> <span class="o">=</span> <span class="n">get_n_muhap_per_core</span><span class="p">(</span><span class="n">pv_core</span><span class="p">,</span> <span class="n">n_muhap_total</span><span class="p">)</span>
<span class="n">greedy_muhap_list</span> <span class="o">=</span> <span class="n">get_greedy_muhap_list</span><span class="p">(</span><span class="n">windows_subset</span><span class="p">,</span> <span class="n">pv_core</span><span class="p">,</span> <span class="n">n_muhap_per_core</span><span class="p">,</span> <span class="mi">10_000</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s plot our greedy microhaplotype marker selection.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_markers</span><span class="p">(</span><span class="n">windows_subset</span><span class="p">,</span><span class="n">greedy_muhap_list</span><span class="p">,</span> <span class="n">pv_regions</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span><span class="s2">&quot;chromStart&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/8e01887d788ef94261fe057d67b0ebaf323c4032c2da3fe297ec881906373698.png" src="_images/8e01887d788ef94261fe057d67b0ebaf323c4032c2da3fe297ec881906373698.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_panel_stats</span><span class="p">(</span><span class="n">windows_subset</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">greedy_muhap_list</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/06f1a2c6c3b115289ebbf300ec883df3f1c6bf2c7077bbc7ab8e8af3e5b96cdf.png" src="_images/06f1a2c6c3b115289ebbf300ec883df3f1c6bf2c7077bbc7ab8e8af3e5b96cdf.png" />
</div>
</div>
<p>We can see that for a marker set of 100, when the markers are prioritised while optimising on informativeness criteria first, and spacing second, there are large gaps between markers, and also regions of the genome where markers clump together (likely higher-diversity regions). If we change the minimum distance (for example to 75,000), we can space them more evenly.</p>
</section>
<section id="evenly-spaced">
<span id="id5"></span><h3>Evenly spaced<a class="headerlink" href="#evenly-spaced" title="Link to this heading">#</a></h3>
<p>A different heuristic approach is to optimise based on spacing as a first priority, and informativeness (in this case heterozygosity) as a second priority. This approach decides the minimum spacing desired between markers (default=1000) and then divides each core region into equal sized segments (proportional to the number of markers required per core region), leaving the desired minimum distance gap in between the segments. Then, a sliding window is moved through each segment, identifying the highest heterozygosity marker window within it.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_max_idx</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">df</span><span class="o">.</span><span class="n">idxmax</span><span class="p">()</span>

<span class="k">def</span> <span class="nf">get_even_muhap_list</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">core</span><span class="p">,</span> <span class="n">n_muhap_per_core</span><span class="p">,</span> <span class="n">spacing</span><span class="o">=</span><span class="mi">1_000</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

    <span class="n">muhap_list</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Iterate through the core regions one by one</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">row</span><span class="p">),</span> <span class="n">n_core</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">core</span><span class="o">.</span><span class="n">iterrows</span><span class="p">(),</span> <span class="n">n_muhap_per_core</span><span class="p">):</span>
        <span class="c1"># Select all windows in the current region and sort them</span>
        <span class="n">core_windows</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">])</span> <span class="o">&amp;</span>
                           <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;window_start&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">])</span> <span class="o">&amp;</span>
                           <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;window_end&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromEnd&quot;</span><span class="p">])</span> <span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">core_windows</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;window_start&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        
        <span class="n">local_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="n">n_core</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span> <span class="c1"># Check that the current region requires at least 1 marker</span>
            <span class="c1"># Calculate the size of the sliding window based on the size of the region, the number of markers,</span>
            <span class="c1"># and the required minimum buffer spacing between them</span>
            <span class="n">win_size</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span><span class="o">/</span><span class="n">n_core</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="n">spacing</span><span class="p">])</span>
            <span class="c1"># Run a sliding window in the region and for each window get the index of the highest heterozygosity marker</span>
            <span class="n">local_list</span><span class="p">,</span> <span class="n">a</span><span class="p">,</span> <span class="n">NN</span> <span class="o">=</span> <span class="n">allel</span><span class="o">.</span><span class="n">windowed_statistic</span><span class="p">(</span><span class="n">core_windows</span><span class="p">[</span><span class="s2">&quot;midpoint&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int&quot;</span><span class="p">),</span> <span class="n">core_windows</span><span class="p">[</span><span class="s2">&quot;het&quot;</span><span class="p">],</span> <span class="n">get_max_idx</span><span class="p">,</span>
                                                         <span class="n">size</span><span class="o">=</span><span class="n">win_size</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span><span class="n">win_size</span><span class="o">+</span><span class="n">spacing</span><span class="p">,</span>
                                                         <span class="n">start</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">],</span> <span class="n">stop</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromEnd&quot;</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">local_list</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span> <span class="c1"># Check if at least one sliding window didn&#39;t find any marker</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Only </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">NN</span><span class="p">)</span><span class="si">}</span><span class="s1"> out of </span><span class="si">{</span><span class="n">n_core</span><span class="si">}</span><span class="s1"> windows found in </span><span class="se">\</span>
<span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;chromEnd&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;L:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">, W:</span><span class="si">{</span><span class="n">win_size</span><span class="si">}</span><span class="s1">, N:</span><span class="si">{</span><span class="n">n_core</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">NN</span><span class="p">)</span>
                <span class="n">local_list</span> <span class="o">=</span> <span class="n">local_list</span><span class="p">[</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">local_list</span><span class="p">)]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int&quot;</span><span class="p">)</span>

        <span class="n">muhap_list</span> <span class="o">=</span> <span class="p">[</span><span class="o">*</span><span class="n">muhap_list</span><span class="p">,</span> <span class="o">*</span><span class="n">local_list</span><span class="p">]</span>
    
    <span class="k">return</span> <span class="n">muhap_list</span>
</pre></div>
</div>
</div>
</div>
<p>Now, we will once again choose which marker windows we would like (variant counts between 3 and 10).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">windows_subset</span> <span class="o">=</span> <span class="n">windows_df</span><span class="p">[</span><span class="n">windows_df</span><span class="p">[</span><span class="s2">&quot;variant_counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">between</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">10</span><span class="p">)]</span>
</pre></div>
</div>
</div>
</div>
<p>We will similarly require 100 markers to be selected, and a minimum distance of 25,000</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_muhap_total</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">n_muhap_per_core</span> <span class="o">=</span> <span class="n">get_n_muhap_per_core</span><span class="p">(</span><span class="n">pv_core</span><span class="p">,</span> <span class="n">n_muhap_total</span><span class="p">)</span>
<span class="n">even_muhap_list</span> <span class="o">=</span> <span class="n">get_even_muhap_list</span><span class="p">(</span><span class="n">windows_subset</span><span class="p">,</span> <span class="n">pv_core</span><span class="p">,</span> <span class="n">n_muhap_per_core</span><span class="p">,</span> <span class="mi">25_000</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">even_muhap_list</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Only 2 out of 3 windows found in PvP01_01_v1:116542-677963
Only 0 out of 1 windows found in PvP01_04_v1:566928-685686
Only 5 out of 6 windows found in PvP01_07_v1:60454-1256419
Only 4 out of 5 windows found in PvP01_08_v1:28424-1132375
Only 3 out of 4 windows found in PvP01_10_v1:140990-1011339
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>95
</pre></div>
</div>
</div>
</div>
<p>Here, we can see that if we optimise on spacing first, we cannot get the desired number of markers (100), only 95 are returned. We can adjust the spacing and number of markers until we can select the desired number.</p>
<p>Let’s plot our <strong>evenly spaced</strong> microhaplotype selection.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_markers</span><span class="p">(</span><span class="n">windows_subset</span><span class="p">,</span><span class="n">even_muhap_list</span><span class="p">,</span> <span class="n">pv_regions</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span><span class="s2">&quot;chromStart&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/5b3a57d7a3a940f2f86421330c253684005e61fe7caa6836f14d21c98fa7c042.png" src="_images/5b3a57d7a3a940f2f86421330c253684005e61fe7caa6836f14d21c98fa7c042.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_panel_stats</span><span class="p">(</span><span class="n">windows_subset</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">even_muhap_list</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/c7acc7d62784a16c29ec87ea4a9cc8f37820d193bd5e853f5fdba03806f655ae.png" src="_images/c7acc7d62784a16c29ec87ea4a9cc8f37820d193bd5e853f5fdba03806f655ae.png" />
</div>
</div>
<p>As we can see, there is a definite improvement in spacing of markers, however, a drawback to this approach is that we could be ignoring good microhaplotype windows with high informativeness by insisting that the spacing be consistent between markers.</p>
<p>There are other, more sophisticated (but computationally demanding) approaches to solve this problem, like <a class="reference external" href="https://en.wikipedia.org/wiki/Dynamic_programming">dynamic programming</a> which can effectively optimise multiple criteria simultaneously (for example, spacing and heterozygosity). This general framework provides a way to get started and experiment different way to select a candidate set of markers. As mentioned at the beginning, regardless of the bioinformatic method used for the candidate selection, it is very likely that a final manual curation is required to completely satisfy, for example, molecular constraints of the specific platform used.</p>
</section>
</section>
<section id="worked-example">
<span id="id6"></span><h2>Worked example<a class="headerlink" href="#worked-example" title="Link to this heading">#</a></h2>
<p>In this last example we run an example end-to-end, from the design of the panel to the preparation for its evaluation using <a class="reference external" href="https://github.com/aimeertaylor/paneljudge">paneljudge</a>. To simplify the process, we start by defining a number of accessory functions.</p>
<p>First, we will use the function below to map variants to windows they belong to and return the position (absolute and relative) of all the variants it containes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">win_to_variants</span><span class="p">(</span><span class="n">windows</span><span class="p">,</span> <span class="n">variants</span><span class="p">):</span>

    <span class="c1"># Intialise output dataframe</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">,</span><span class="s2">&quot;length&quot;</span><span class="p">,</span><span class="s2">&quot;number&quot;</span><span class="p">,</span><span class="s2">&quot;variants&quot;</span><span class="p">])</span>
    
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">win</span> <span class="ow">in</span> <span class="n">windows</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span> <span class="c1"># For each selected window</span>
        <span class="c1"># Find all variants that fall within the window</span>
        <span class="n">var_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">((</span><span class="n">variants</span><span class="p">[</span><span class="s2">&quot;variant_chrom&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">win</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">])</span> <span class="o">&amp;</span>
                           <span class="p">(</span><span class="n">variants</span><span class="p">[</span><span class="s2">&quot;variant_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">between</span><span class="p">(</span><span class="n">win</span><span class="p">[</span><span class="s2">&quot;window_start&quot;</span><span class="p">],</span><span class="n">win</span><span class="p">[</span><span class="s2">&quot;window_end&quot;</span><span class="p">])))[</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">n_var</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">var_idx</span><span class="p">)</span>
        <span class="c1"># Throw an error in case the number of variants found doesn&#39;t match the expectation</span>
        <span class="k">assert</span> <span class="n">n_var</span><span class="o">==</span><span class="n">win</span><span class="p">[</span><span class="s2">&quot;variant_counts&quot;</span><span class="p">],</span> <span class="s2">&quot;Unexpected number of variants - are all files in sync?&quot;</span>
        
        <span class="c1"># The microhaplotype unique ID will have the following format:</span>
        <span class="c1"># &lt;crhomosome&gt;:&lt;position of variant #1&gt;:[&lt;position of variant #2 relative to #1&gt;, ...]</span>
        <span class="n">relative_pos</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">var_idx</span><span class="p">[</span><span class="mi">1</span><span class="p">:]][</span><span class="s2">&quot;variant_position&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">variants</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">var_idx</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="s2">&quot;variant_position&quot;</span><span class="p">]</span>
        <span class="n">mhap_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="s2">&quot;:&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">variant_coords</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">var_idx</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">relative_pos</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span>
        
        <span class="n">mhap_start</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">var_idx</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="s2">&quot;variant_position&quot;</span><span class="p">]</span>
        <span class="n">mhap_end</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">var_idx</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]][</span><span class="s2">&quot;variant_position&quot;</span><span class="p">]</span>
        <span class="n">mhap_length</span> <span class="o">=</span> <span class="n">mhap_end</span> <span class="o">-</span> <span class="n">mhap_start</span>

        <span class="c1"># For convenience, also include the absolute position of all variants in the window</span>
        <span class="n">var_list</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">var_idx</span><span class="p">:</span>
            <span class="n">var_list</span> <span class="o">+=</span> <span class="p">[</span><span class="s2">&quot;:&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">variant_coords</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">v</span><span class="p">])]</span>

        <span class="c1"># Add a new row to the dataframe. The index matches the one in the &#39;windows&#39; dataframe, useful for merging</span>
        <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;id&quot;</span><span class="p">:</span><span class="n">mhap_id</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">:</span><span class="n">mhap_start</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">:</span><span class="n">mhap_end</span><span class="p">,</span> <span class="s2">&quot;length&quot;</span><span class="p">:</span><span class="n">mhap_length</span><span class="p">,</span>
                       <span class="s2">&quot;number&quot;</span><span class="p">:</span><span class="n">n_var</span><span class="p">,</span> <span class="s2">&quot;variants&quot;</span><span class="p">:</span><span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">var_list</span><span class="p">)}</span>

    <span class="k">return</span> <span class="n">df</span>
</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">paneljudge</span></code> treats each microhaplotype as a single variant rather than a collection of individual ones. The microhaplotype frequencies are a required input to run the simulations and the function below calculates them. For a given microhaplotype and the variants it contains, the function extracts the sample genotypes and transforms them into <em>haplotypes</em> (i.e. the concatenation of the allele state of each variant). For example, if a microhaplotype contains two biallelic SNPs, there are up to four theoretical haplotypes possible (ab, Ab, aB, AB) corresponding to as many microhaplotype alleles.</p>
<p>Before calculating frequencies, we need to define a strategy to handle “heterozygous” (i.e. the sample contains a collection of parasites with different alleles in that position) and “missing” (i.e. not enough reads to confidently output an allele) genotype calls. The strict sample selection criteria mean that those calls are overall very rare, to the point that different approaches have very limited impact on the final results. Because of that and for the sake of simplicity, we decided to exclude any haplotype containing missing calls and randomly select one of the alleles in heterozygous calls (done under the hood by the <code class="docutils literal notranslate"><span class="pre">scikit-allel</span></code> function <code class="docutils literal notranslate"><span class="pre">haploidify_samples()</span></code>). We explicitly define the random seed for reproducibility purposes.</p>
<p>This is all handled within the function below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">haplotype_freq</span><span class="p">(</span><span class="n">amplicon</span><span class="p">,</span> <span class="n">variants</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">250523</span><span class="p">):</span>
    
    <span class="n">muhap</span> <span class="o">=</span> <span class="n">amplicon</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
    <span class="n">va</span><span class="o">=</span><span class="n">variants</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="n">muhap</span><span class="p">)</span>
    <span class="n">gt</span> <span class="o">=</span> <span class="n">allel</span><span class="o">.</span><span class="n">GenotypeArray</span><span class="p">(</span><span class="n">va</span><span class="p">[</span><span class="s2">&quot;call_genotype&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
    
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">haplo</span> <span class="o">=</span> <span class="n">gt</span><span class="o">.</span><span class="n">haploidify_samples</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">haplo_str</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">haplo</span><span class="o">.</span><span class="n">transpose</span><span class="p">()]</span>
    <span class="n">haplo_non_missing</span> <span class="o">=</span> <span class="p">[</span><span class="n">h</span> <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">haplo_str</span> <span class="k">if</span> <span class="s2">&quot;-&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">h</span><span class="p">]</span>

    <span class="n">values</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">haplo_non_missing</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">freqs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">counts</span><span class="o">/</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">counts</span><span class="p">))[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">freqs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">freqs</span><span class="p">,</span><span class="n">decimals</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">freqs</span>
</pre></div>
</div>
</div>
</div>
<p>To create our example panel, we begin by selecting only the windows that satisfy our minimum criteria, in this example between 3 and 10 variants and heterozygosity of at least 0.6.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">windows_subset</span> <span class="o">=</span> <span class="n">windows_df</span><span class="p">[(</span><span class="n">windows_df</span><span class="p">[</span><span class="s2">&quot;variant_counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">between</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span><span class="o">&amp;</span><span class="p">(</span><span class="n">windows_df</span><span class="p">[</span><span class="s2">&quot;het&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mf">0.6</span><span class="p">)]</span>
<span class="nb">len</span><span class="p">(</span><span class="n">windows_subset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1110
</pre></div>
</div>
</div>
</div>
<p>We then select 50 evenly-spaced windows and visualise some basic summary statistics for the panel using the functions defined above:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_muhap_per_core</span> <span class="o">=</span> <span class="n">get_n_muhap_per_core</span><span class="p">(</span><span class="n">pv_core</span><span class="p">,</span> <span class="mi">50</span><span class="p">)</span>
<span class="n">muhap_list</span> <span class="o">=</span> <span class="n">get_even_muhap_list</span><span class="p">(</span><span class="n">windows_subset</span><span class="p">,</span> <span class="n">pv_core</span><span class="p">,</span> <span class="n">n_muhap_per_core</span><span class="p">,</span> <span class="n">spacing</span><span class="o">=</span><span class="mi">10_000</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_markers</span><span class="p">(</span><span class="n">windows_subset</span><span class="p">,</span><span class="n">muhap_list</span><span class="p">,</span> <span class="n">pv_regions</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span><span class="s2">&quot;chromStart&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/17e0c0f7739ade52e07ffea55f897f6f21fe882d06036dd46d4000a4464ffc01.png" src="_images/17e0c0f7739ade52e07ffea55f897f6f21fe882d06036dd46d4000a4464ffc01.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_panel_stats</span><span class="p">(</span><span class="n">windows_subset</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">muhap_list</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/a3702f6c7990a754def6ee4ae5f2924609754d0a1146317274b4b8ff9309686b.png" src="_images/a3702f6c7990a754def6ee4ae5f2924609754d0a1146317274b4b8ff9309686b.png" />
</div>
</div>
<p>At this point we could experiment with other selection algorithms, different parameters, introduce some degree of manual curation, etc. Once happy, we can proceed to extracting the actual list of variants:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">panel</span><span class="o">=</span><span class="n">win_to_variants</span><span class="p">(</span><span class="n">windows_subset</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">muhap_list</span><span class="p">],</span> <span class="n">variant_coords</span><span class="p">)</span>
<span class="n">panel</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>id</th>
      <th>start</th>
      <th>end</th>
      <th>length</th>
      <th>number</th>
      <th>variants</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>175</th>
      <td>PvP01_01_v1:395454:27,40,55,67,154</td>
      <td>395454</td>
      <td>395608</td>
      <td>154</td>
      <td>6</td>
      <td>PvP01_01_v1:395454;PvP01_01_v1:395481;PvP01_01...</td>
    </tr>
    <tr>
      <th>467</th>
      <td>PvP01_01_v1:772094:42,60,64,65,117</td>
      <td>772094</td>
      <td>772211</td>
      <td>117</td>
      <td>6</td>
      <td>PvP01_01_v1:772094;PvP01_01_v1:772136;PvP01_01...</td>
    </tr>
    <tr>
      <th>742</th>
      <td>PvP01_02_v1:327553:35,63,129,186</td>
      <td>327553</td>
      <td>327739</td>
      <td>186</td>
      <td>5</td>
      <td>PvP01_02_v1:327553;PvP01_02_v1:327588;PvP01_02...</td>
    </tr>
    <tr>
      <th>1247</th>
      <td>PvP01_03_v1:178938:83,84,152,156</td>
      <td>178938</td>
      <td>179094</td>
      <td>156</td>
      <td>5</td>
      <td>PvP01_03_v1:178938;PvP01_03_v1:179021;PvP01_03...</td>
    </tr>
    <tr>
      <th>1760</th>
      <td>PvP01_03_v1:831863:12,111,135</td>
      <td>831863</td>
      <td>831998</td>
      <td>135</td>
      <td>4</td>
      <td>PvP01_03_v1:831863;PvP01_03_v1:831875;PvP01_03...</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">panel</span><span class="p">[</span><span class="s2">&quot;number&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>285
</pre></div>
</div>
</div>
</div>
<p>There are in total 285 variants across the 50 selected windows.</p>
<p>We are now ready to prepare the input files for <code class="docutils literal notranslate"><span class="pre">paneljudge</span></code>. In order to do so, we need to decide which samples (or population) to use to calculate the haplotype frequencies. As an example, let’s use samples from the eastern part of Southeast Asia (see the Pv4 paper for details on how populations are defined based on geographic and genetic homogeneity):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filter the samples based on the criteria above</span>
<span class="n">loc_filtered_samples</span> <span class="o">=</span> <span class="p">(</span>
    <span class="p">(</span><span class="n">sample_metadata</span><span class="p">[</span><span class="s2">&quot;Population&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ESEA&quot;</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">sample_metadata</span><span class="p">[</span><span class="s2">&quot;Fws&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.95</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">sample_metadata</span><span class="p">[</span><span class="s2">&quot;</span><span class="si">% c</span><span class="s2">allable&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">50</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">sample_metadata</span><span class="p">[</span><span class="s2">&quot;Exclusion reason&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Analysis_set&quot;</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">subset_metadata</span> <span class="o">=</span> <span class="n">sample_metadata</span><span class="p">[</span><span class="n">loc_filtered_samples</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">subset_metadata</span><span class="p">))</span>
<span class="n">variant_dataset_filtered_samples</span> <span class="o">=</span> <span class="n">variant_dataset</span><span class="o">.</span><span class="n">isel</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="n">loc_filtered_samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>161
</pre></div>
</div>
</div>
</div>
<p>The next part speeds-up the computation by reducing the genomics data to the 161 samples and 285 variants needed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variables_to_remove</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">variant_dataset_filtered_samples</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="c1"># Add all variable names to remove list...</span>
<span class="n">variables_to_remove</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s2">&quot;call_genotype&quot;</span><span class="p">)</span> <span class="c1"># ...but we need to keep the genotypes</span>
<span class="n">reduced_variant_set</span> <span class="o">=</span> <span class="n">variant_dataset_filtered_samples</span><span class="o">.</span><span class="n">drop_vars</span><span class="p">(</span><span class="n">variables_to_remove</span><span class="p">)</span> <span class="c1"># Reduce the dataset</span>

<span class="c1"># Pre-load the data to avoid continuous query to the cloud. Not strictly speaking necessary.</span>
<span class="n">ids</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">c</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span><span class="n">p</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">variant_dataset_filtered_samples</span><span class="o">.</span><span class="n">coords</span><span class="p">[</span><span class="s2">&quot;variant_chrom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                                 <span class="n">variant_dataset_filtered_samples</span><span class="o">.</span><span class="n">coords</span><span class="p">[</span><span class="s2">&quot;variant_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)]</span>
<span class="n">reduced_variant_set</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;variants&quot;</span><span class="p">,</span><span class="n">ids</span><span class="p">)</span>
<span class="n">all_variants</span> <span class="o">=</span> <span class="n">reduced_variant_set</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="s2">&quot;id&quot;</span><span class="p">)</span>
<span class="n">all_variants</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewBox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewBox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body[data-theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block !important;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-index-preview {
  grid-column: 2 / 5;
  color: var(--xr-font-color2);
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data,
.xr-index-data-in:checked ~ .xr-index-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-index-name div,
.xr-index-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2,
.xr-no-icon {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;
Dimensions:           (variants: 4571056, samples: 161, ploidy: 2)
Coordinates:
    variant_position  (variants) int32 dask.array&lt;chunksize=(65536,), meta=np.ndarray&gt;
    variant_chrom     (variants) object dask.array&lt;chunksize=(65536,), meta=np.ndarray&gt;
    sample_id         (samples) object dask.array&lt;chunksize=(161,), meta=np.ndarray&gt;
  * variants          (variants) &lt;U33 &#x27;PvP01_01_v1:23&#x27; ... &#x27;Transfer.PvP01_00...
Dimensions without coordinates: samples, ploidy
Data variables:
    call_genotype     (variants, samples, ploidy) int8 dask.array&lt;chunksize=(65536, 2, 2), meta=np.ndarray&gt;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-fc3be3aa-8ec5-4736-8509-63acf167d0e4' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-fc3be3aa-8ec5-4736-8509-63acf167d0e4' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span class='xr-has-index'>variants</span>: 4571056</li><li><span>samples</span>: 161</li><li><span>ploidy</span>: 2</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-57db181a-7357-43a4-9936-b5d2b8661e82' class='xr-section-summary-in' type='checkbox'  checked><label for='section-57db181a-7357-43a4-9936-b5d2b8661e82' class='xr-section-summary' >Coordinates: <span>(4)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>variant_position</span></div><div class='xr-var-dims'>(variants)</div><div class='xr-var-dtype'>int32</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(65536,), meta=np.ndarray&gt;</div><input id='attrs-0d6bdc10-18a8-43b5-bfec-754c37e73b3b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0d6bdc10-18a8-43b5-bfec-754c37e73b3b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4d224cde-de95-4a91-a57c-2a1fa6b261ec' class='xr-var-data-in' type='checkbox'><label for='data-4d224cde-de95-4a91-a57c-2a1fa6b261ec' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
    <tr>
        <td>
            <table style="border-collapse: collapse;">
                <thead>
                    <tr>
                        <td> </td>
                        <th> Array </th>
                        <th> Chunk </th>
                    </tr>
                </thead>
                <tbody>
                    
                    <tr>
                        <th> Bytes </th>
                        <td> 17.44 MiB </td>
                        <td> 256.00 kiB </td>
                    </tr>
                    
                    <tr>
                        <th> Shape </th>
                        <td> (4571056,) </td>
                        <td> (65536,) </td>
                    </tr>
                    <tr>
                        <th> Dask graph </th>
                        <td colspan="2"> 70 chunks in 1 graph layer </td>
                    </tr>
                    <tr>
                        <th> Data type </th>
                        <td colspan="2"> int32 numpy.ndarray </td>
                    </tr>
                </tbody>
            </table>
        </td>
        <td>
        <svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="5" y1="0" x2="5" y2="25" />
  <line x1="12" y1="0" x2="12" y2="25" />
  <line x1="18" y1="0" x2="18" y2="25" />
  <line x1="24" y1="0" x2="24" y2="25" />
  <line x1="30" y1="0" x2="30" y2="25" />
  <line x1="37" y1="0" x2="37" y2="25" />
  <line x1="43" y1="0" x2="43" y2="25" />
  <line x1="49" y1="0" x2="49" y2="25" />
  <line x1="56" y1="0" x2="56" y2="25" />
  <line x1="61" y1="0" x2="61" y2="25" />
  <line x1="68" y1="0" x2="68" y2="25" />
  <line x1="75" y1="0" x2="75" y2="25" />
  <line x1="80" y1="0" x2="80" y2="25" />
  <line x1="87" y1="0" x2="87" y2="25" />
  <line x1="94" y1="0" x2="94" y2="25" />
  <line x1="99" y1="0" x2="99" y2="25" />
  <line x1="106" y1="0" x2="106" y2="25" />
  <line x1="113" y1="0" x2="113" y2="25" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >4571056</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
        </td>
    </tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>variant_chrom</span></div><div class='xr-var-dims'>(variants)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(65536,), meta=np.ndarray&gt;</div><input id='attrs-f955d648-d597-4d8a-b4b8-6de65689369c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f955d648-d597-4d8a-b4b8-6de65689369c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-437dcd0c-9e75-433b-beff-ccbe8a1667e6' class='xr-var-data-in' type='checkbox'><label for='data-437dcd0c-9e75-433b-beff-ccbe8a1667e6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
    <tr>
        <td>
            <table style="border-collapse: collapse;">
                <thead>
                    <tr>
                        <td> </td>
                        <th> Array </th>
                        <th> Chunk </th>
                    </tr>
                </thead>
                <tbody>
                    
                    <tr>
                        <th> Bytes </th>
                        <td> 34.87 MiB </td>
                        <td> 512.00 kiB </td>
                    </tr>
                    
                    <tr>
                        <th> Shape </th>
                        <td> (4571056,) </td>
                        <td> (65536,) </td>
                    </tr>
                    <tr>
                        <th> Dask graph </th>
                        <td colspan="2"> 70 chunks in 1 graph layer </td>
                    </tr>
                    <tr>
                        <th> Data type </th>
                        <td colspan="2"> object numpy.ndarray </td>
                    </tr>
                </tbody>
            </table>
        </td>
        <td>
        <svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="5" y1="0" x2="5" y2="25" />
  <line x1="12" y1="0" x2="12" y2="25" />
  <line x1="18" y1="0" x2="18" y2="25" />
  <line x1="24" y1="0" x2="24" y2="25" />
  <line x1="30" y1="0" x2="30" y2="25" />
  <line x1="37" y1="0" x2="37" y2="25" />
  <line x1="43" y1="0" x2="43" y2="25" />
  <line x1="49" y1="0" x2="49" y2="25" />
  <line x1="56" y1="0" x2="56" y2="25" />
  <line x1="61" y1="0" x2="61" y2="25" />
  <line x1="68" y1="0" x2="68" y2="25" />
  <line x1="75" y1="0" x2="75" y2="25" />
  <line x1="80" y1="0" x2="80" y2="25" />
  <line x1="87" y1="0" x2="87" y2="25" />
  <line x1="94" y1="0" x2="94" y2="25" />
  <line x1="99" y1="0" x2="99" y2="25" />
  <line x1="106" y1="0" x2="106" y2="25" />
  <line x1="113" y1="0" x2="113" y2="25" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >4571056</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
        </td>
    </tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>sample_id</span></div><div class='xr-var-dims'>(samples)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(161,), meta=np.ndarray&gt;</div><input id='attrs-5d00e7ac-947e-4b09-ad26-529fa047d15a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5d00e7ac-947e-4b09-ad26-529fa047d15a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5fada442-3bb5-4132-9fd7-f96f144de1bf' class='xr-var-data-in' type='checkbox'><label for='data-5fada442-3bb5-4132-9fd7-f96f144de1bf' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
    <tr>
        <td>
            <table style="border-collapse: collapse;">
                <thead>
                    <tr>
                        <td> </td>
                        <th> Array </th>
                        <th> Chunk </th>
                    </tr>
                </thead>
                <tbody>
                    
                    <tr>
                        <th> Bytes </th>
                        <td> 1.26 kiB </td>
                        <td> 1.26 kiB </td>
                    </tr>
                    
                    <tr>
                        <th> Shape </th>
                        <td> (161,) </td>
                        <td> (161,) </td>
                    </tr>
                    <tr>
                        <th> Dask graph </th>
                        <td colspan="2"> 1 chunks in 2 graph layers </td>
                    </tr>
                    <tr>
                        <th> Data type </th>
                        <td colspan="2"> object numpy.ndarray </td>
                    </tr>
                </tbody>
            </table>
        </td>
        <td>
        <svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >161</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
        </td>
    </tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>variants</span></div><div class='xr-var-dims'>(variants)</div><div class='xr-var-dtype'>&lt;U33</div><div class='xr-var-preview xr-preview'>&#x27;PvP01_01_v1:23&#x27; ... &#x27;Transfer.P...</div><input id='attrs-b99c33ad-0db1-4c78-9743-34513f109ae9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b99c33ad-0db1-4c78-9743-34513f109ae9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0aa6bcca-5401-4444-9df4-45cceff5a7b3' class='xr-var-data-in' type='checkbox'><label for='data-0aa6bcca-5401-4444-9df4-45cceff5a7b3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;PvP01_01_v1:23&#x27;, &#x27;PvP01_01_v1:26&#x27;, &#x27;PvP01_01_v1:27&#x27;, ...,
       &#x27;Transfer.PvP01_00_99.final:3769&#x27;, &#x27;Transfer.PvP01_00_99.final:3771&#x27;,
       &#x27;Transfer.PvP01_00_99.final:3774&#x27;], dtype=&#x27;&lt;U33&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-474b74df-2340-4f89-b995-3415362ca4c9' class='xr-section-summary-in' type='checkbox'  checked><label for='section-474b74df-2340-4f89-b995-3415362ca4c9' class='xr-section-summary' >Data variables: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>call_genotype</span></div><div class='xr-var-dims'>(variants, samples, ploidy)</div><div class='xr-var-dtype'>int8</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(65536, 2, 2), meta=np.ndarray&gt;</div><input id='attrs-0323ff2f-b123-4d5f-8be8-de47dbe0ca26' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0323ff2f-b123-4d5f-8be8-de47dbe0ca26' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-48ffd4e2-1324-4f6e-b9b8-73626cc59dfd' class='xr-var-data-in' type='checkbox'><label for='data-48ffd4e2-1324-4f6e-b9b8-73626cc59dfd' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
    <tr>
        <td>
            <table style="border-collapse: collapse;">
                <thead>
                    <tr>
                        <td> </td>
                        <th> Array </th>
                        <th> Chunk </th>
                    </tr>
                </thead>
                <tbody>
                    
                    <tr>
                        <th> Bytes </th>
                        <td> 1.37 GiB </td>
                        <td> 3.75 MiB </td>
                    </tr>
                    
                    <tr>
                        <th> Shape </th>
                        <td> (4571056, 161, 2) </td>
                        <td> (65536, 30, 2) </td>
                    </tr>
                    <tr>
                        <th> Dask graph </th>
                        <td colspan="2"> 1120 chunks in 2 graph layers </td>
                    </tr>
                    <tr>
                        <th> Data type </th>
                        <td colspan="2"> int8 numpy.ndarray </td>
                    </tr>
                </tbody>
            </table>
        </td>
        <td>
        <svg width="156" height="146" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="10" y1="0" x2="80" y2="70" />
  <line x1="10" y1="5" x2="80" y2="75" />
  <line x1="10" y1="5" x2="80" y2="76" />
  <line x1="10" y1="5" x2="80" y2="76" />
  <line x1="10" y1="8" x2="80" y2="78" />
  <line x1="10" y1="10" x2="80" y2="80" />
  <line x1="10" y1="11" x2="80" y2="81" />
  <line x1="10" y1="11" x2="80" y2="81" />
  <line x1="10" y1="11" x2="80" y2="82" />
  <line x1="10" y1="14" x2="80" y2="85" />
  <line x1="10" y1="16" x2="80" y2="87" />
  <line x1="10" y1="17" x2="80" y2="88" />
  <line x1="10" y1="19" x2="80" y2="89" />
  <line x1="10" y1="22" x2="80" y2="92" />
  <line x1="10" y1="23" x2="80" y2="94" />
  <line x1="10" y1="25" x2="80" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="10" y2="25" style="stroke-width:2" />
  <line x1="13" y1="3" x2="13" y2="28" />
  <line x1="17" y1="7" x2="17" y2="32" />
  <line x1="21" y1="11" x2="21" y2="36" />
  <line x1="24" y1="14" x2="24" y2="39" />
  <line x1="28" y1="18" x2="28" y2="43" />
  <line x1="32" y1="22" x2="32" y2="47" />
  <line x1="35" y1="25" x2="35" y2="50" />
  <line x1="39" y1="29" x2="39" y2="54" />
  <line x1="43" y1="33" x2="43" y2="58" />
  <line x1="46" y1="36" x2="46" y2="61" />
  <line x1="50" y1="40" x2="50" y2="65" />
  <line x1="54" y1="44" x2="54" y2="69" />
  <line x1="57" y1="47" x2="57" y2="72" />
  <line x1="61" y1="51" x2="61" y2="77" />
  <line x1="65" y1="55" x2="65" y2="81" />
  <line x1="68" y1="58" x2="68" y2="84" />
  <line x1="72" y1="62" x2="72" y2="88" />
  <line x1="76" y1="66" x2="76" y2="92" />
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 80.58823529411765,70.58823529411765 80.58823529411765,96.00085180870013 10.0,25.412616514582485" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="35" y2="0" style="stroke-width:2" />
  <line x1="13" y1="3" x2="38" y2="3" />
  <line x1="17" y1="7" x2="42" y2="7" />
  <line x1="21" y1="11" x2="46" y2="11" />
  <line x1="24" y1="14" x2="49" y2="14" />
  <line x1="28" y1="18" x2="53" y2="18" />
  <line x1="32" y1="22" x2="57" y2="22" />
  <line x1="35" y1="25" x2="60" y2="25" />
  <line x1="39" y1="29" x2="64" y2="29" />
  <line x1="43" y1="33" x2="68" y2="33" />
  <line x1="46" y1="36" x2="71" y2="36" />
  <line x1="50" y1="40" x2="75" y2="40" />
  <line x1="54" y1="44" x2="79" y2="44" />
  <line x1="57" y1="47" x2="82" y2="47" />
  <line x1="61" y1="51" x2="87" y2="51" />
  <line x1="65" y1="55" x2="91" y2="55" />
  <line x1="68" y1="58" x2="94" y2="58" />
  <line x1="72" y1="62" x2="98" y2="62" />
  <line x1="76" y1="66" x2="102" y2="66" />
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="35" y1="0" x2="106" y2="70" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 35.41261651458248,0.0 106.00085180870013,70.58823529411765 80.58823529411765,70.58823529411765" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />
  <line x1="80" y1="70" x2="106" y2="70" />
  <line x1="80" y1="75" x2="106" y2="75" />
  <line x1="80" y1="76" x2="106" y2="76" />
  <line x1="80" y1="76" x2="106" y2="76" />
  <line x1="80" y1="78" x2="106" y2="78" />
  <line x1="80" y1="80" x2="106" y2="80" />
  <line x1="80" y1="81" x2="106" y2="81" />
  <line x1="80" y1="81" x2="106" y2="81" />
  <line x1="80" y1="82" x2="106" y2="82" />
  <line x1="80" y1="85" x2="106" y2="85" />
  <line x1="80" y1="87" x2="106" y2="87" />
  <line x1="80" y1="88" x2="106" y2="88" />
  <line x1="80" y1="89" x2="106" y2="89" />
  <line x1="80" y1="92" x2="106" y2="92" />
  <line x1="80" y1="94" x2="106" y2="94" />
  <line x1="80" y1="96" x2="106" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />
  <line x1="106" y1="70" x2="106" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="80.58823529411765,70.58823529411765 106.00085180870013,70.58823529411765 106.00085180870013,96.00085180870013 80.58823529411765,96.00085180870013" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="93.294544" y="116.000852" font-size="1.0rem" font-weight="100" text-anchor="middle" >2</text>
  <text x="126.000852" y="83.294544" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,126.000852,83.294544)">161</text>
  <text x="35.294118" y="80.706734" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(45,35.294118,80.706734)">4571056</text>
</svg>
        </td>
    </tr>
</table></div></li></ul></div></li><li class='xr-section-item'><input id='section-152cf301-5250-475b-9514-f1850ea693ed' class='xr-section-summary-in' type='checkbox'  ><label for='section-152cf301-5250-475b-9514-f1850ea693ed' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>variants</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-37f33a4d-8ad2-4b57-b303-d78b309ab9d5' class='xr-index-data-in' type='checkbox'/><label for='index-37f33a4d-8ad2-4b57-b303-d78b309ab9d5' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;PvP01_01_v1:23&#x27;, &#x27;PvP01_01_v1:26&#x27;, &#x27;PvP01_01_v1:27&#x27;, &#x27;PvP01_01_v1:34&#x27;,
       &#x27;PvP01_01_v1:41&#x27;, &#x27;PvP01_01_v1:42&#x27;, &#x27;PvP01_01_v1:43&#x27;, &#x27;PvP01_01_v1:48&#x27;,
       &#x27;PvP01_01_v1:52&#x27;, &#x27;PvP01_01_v1:54&#x27;,
       ...
       &#x27;Transfer.PvP01_00_99.final:3760&#x27;, &#x27;Transfer.PvP01_00_99.final:3761&#x27;,
       &#x27;Transfer.PvP01_00_99.final:3762&#x27;, &#x27;Transfer.PvP01_00_99.final:3763&#x27;,
       &#x27;Transfer.PvP01_00_99.final:3764&#x27;, &#x27;Transfer.PvP01_00_99.final:3765&#x27;,
       &#x27;Transfer.PvP01_00_99.final:3766&#x27;, &#x27;Transfer.PvP01_00_99.final:3769&#x27;,
       &#x27;Transfer.PvP01_00_99.final:3771&#x27;, &#x27;Transfer.PvP01_00_99.final:3774&#x27;],
      dtype=&#x27;object&#x27;, name=&#x27;variants&#x27;, length=4571056))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-257889b2-d420-445c-8136-b9a569cb6224' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-257889b2-d420-445c-8136-b9a569cb6224' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div></div></div>
</div>
<p>The final step is to create actual input files by going through all microhaplotypes one by one.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">create_pj_files</span><span class="p">(</span><span class="n">panel</span><span class="p">,</span> <span class="n">variants</span><span class="p">):</span>

    <span class="c1"># Initialise the dataframe</span>
    <span class="n">pj</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Amplicon_name&quot;</span><span class="p">,</span><span class="s2">&quot;Chr&quot;</span><span class="p">,</span><span class="s2">&quot;Start&quot;</span><span class="p">,</span><span class="s2">&quot;Stop&quot;</span><span class="p">,</span><span class="s2">&quot;length&quot;</span><span class="p">,</span><span class="s2">&quot;pos&quot;</span><span class="p">,</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span><span class="s2">&quot;distances&quot;</span><span class="p">,</span><span class="s2">&quot;freqs&quot;</span><span class="p">],</span><span class="n">index</span><span class="o">=</span><span class="n">panel</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">])</span>

    <span class="c1"># Interate through each window/microhaplotype</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">amplicon</span> <span class="ow">in</span> <span class="n">panel</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>

        <span class="c1"># Convert chromosome names to number (e.g. &#39;PvP01_07_v1&#39; -&gt; 13)</span>
        <span class="n">chr_name</span> <span class="o">=</span> <span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">chr_num</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">chr_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>

        <span class="c1"># Fill in all fields required by paneljudge</span>
        <span class="n">pj</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]][</span><span class="s2">&quot;Amplicon_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
        <span class="n">pj</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]][</span><span class="s2">&quot;Chr&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">chr_name</span>
        <span class="n">pj</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]][</span><span class="s2">&quot;Start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span>
        <span class="n">pj</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]][</span><span class="s2">&quot;Stop&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
        <span class="n">pj</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]][</span><span class="s2">&quot;length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span><span class="o">-</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span>
        <span class="n">pj</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]][</span><span class="s2">&quot;pos&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">],</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]]))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;int&quot;</span><span class="p">)</span>
        <span class="n">pj</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]][</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">chr_num</span>
        <span class="c1"># This step is the most demanding as it requires access to the genomic data</span>
        <span class="n">pj</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]][</span><span class="s2">&quot;freqs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">haplotype_freq</span><span class="p">(</span><span class="n">amplicon</span><span class="p">[</span><span class="s2">&quot;variants&quot;</span><span class="p">],</span> <span class="n">variants</span><span class="p">)</span>

        <span class="c1"># Finally, calculate the forward intra-markers distance</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pj</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pj</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">pj</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">pj</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">][</span><span class="s2">&quot;chrom&quot;</span><span class="p">]:</span>
                    <span class="n">pj</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;distances&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pj</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">][</span><span class="s2">&quot;pos&quot;</span><span class="p">]</span><span class="o">-</span><span class="n">pj</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;pos&quot;</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">pj</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;distances&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Inf&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">pj</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;distances&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Inf&quot;</span>
    
    <span class="k">return</span> <span class="n">pj</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="n">pj_df</span> <span class="o">=</span> <span class="n">create_pj_files</span><span class="p">(</span><span class="n">panel</span><span class="p">,</span><span class="n">all_variants</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CPU times: user 21.4 s, sys: 2.78 s, total: 24.1 s
Wall time: 1min 18s
</pre></div>
</div>
</div>
</div>
<p>Data are ready to be saved in tsv files to be later processed in R. We need two files, one containing the microhaplotype marker details (genomic position, etc), and one containing the frequency of each allele (per sub-population).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save the markers file</span>
<span class="n">pj_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;freqs&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;../precomputed/markers.tsv&quot;</span><span class="p">,</span> <span class="n">index_label</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1"># paneljudge requires the frequencies to be linearised (one row per microhaplotype, one column per allele)</span>
<span class="c1"># The next lines calculates how many columns are required (max number of alleles across all microhaplotypes)</span>
<span class="n">ff</span> <span class="o">=</span> <span class="n">pj_df</span><span class="p">[</span><span class="s2">&quot;freqs&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
<span class="n">max_alleles</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="n">i</span><span class="o">.</span><span class="n">size</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ff</span><span class="p">])</span>
<span class="c1"># Create and save the frequencies dataframe</span>
<span class="n">pj_freqs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">ff</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="n">pj_df</span><span class="p">[</span><span class="s2">&quot;Amplicon_name&quot;</span><span class="p">],</span>
                                    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Allele.</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">max_alleles</span><span class="p">)])</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">pj_freqs</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;../precomputed/frequencies.tsv&quot;</span><span class="p">,</span> <span class="n">index_label</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">pj_freqs</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Allele.0</th>
      <th>Allele.1</th>
      <th>Allele.2</th>
      <th>Allele.3</th>
      <th>Allele.4</th>
      <th>Allele.5</th>
      <th>Allele.6</th>
      <th>Allele.7</th>
      <th>Allele.8</th>
      <th>Allele.9</th>
      <th>...</th>
      <th>Allele.29</th>
      <th>Allele.30</th>
      <th>Allele.31</th>
      <th>Allele.32</th>
      <th>Allele.33</th>
      <th>Allele.34</th>
      <th>Allele.35</th>
      <th>Allele.36</th>
      <th>Allele.37</th>
      <th>Allele.38</th>
    </tr>
    <tr>
      <th>Amplicon_name</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>PvP01_01_v1:395454:27,40,55,67,154</th>
      <td>0.44720</td>
      <td>0.18634</td>
      <td>0.12422</td>
      <td>0.06211</td>
      <td>0.05590</td>
      <td>0.04969</td>
      <td>0.02484</td>
      <td>0.02484</td>
      <td>0.00621</td>
      <td>0.00621</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>PvP01_01_v1:772094:42,60,64,65,117</th>
      <td>0.51899</td>
      <td>0.25316</td>
      <td>0.05063</td>
      <td>0.03797</td>
      <td>0.03797</td>
      <td>0.02532</td>
      <td>0.02532</td>
      <td>0.02532</td>
      <td>0.01899</td>
      <td>0.00633</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>PvP01_02_v1:327553:35,63,129,186</th>
      <td>0.48447</td>
      <td>0.11801</td>
      <td>0.09317</td>
      <td>0.06832</td>
      <td>0.06211</td>
      <td>0.05590</td>
      <td>0.03106</td>
      <td>0.01863</td>
      <td>0.01242</td>
      <td>0.01242</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>PvP01_03_v1:178938:83,84,152,156</th>
      <td>0.37736</td>
      <td>0.23899</td>
      <td>0.19497</td>
      <td>0.16352</td>
      <td>0.01258</td>
      <td>0.00629</td>
      <td>0.00629</td>
      <td>0.00000</td>
      <td>0.00000</td>
      <td>0.00000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>PvP01_03_v1:831863:12,111,135</th>
      <td>0.31013</td>
      <td>0.22152</td>
      <td>0.13924</td>
      <td>0.11392</td>
      <td>0.05063</td>
      <td>0.04430</td>
      <td>0.03165</td>
      <td>0.02532</td>
      <td>0.01899</td>
      <td>0.01266</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 39 columns</p>
</div></div></div>
</div>
<p>The two files (in this example <code class="docutils literal notranslate"><span class="pre">markers.tsv</span></code> and <code class="docutils literal notranslate"><span class="pre">frequencies.tsv</span></code> in the <code class="docutils literal notranslate"><span class="pre">precomputed</span></code> directory) are now ready to be read in R before running <code class="docutils literal notranslate"><span class="pre">paneljudge</span></code>. The snippet of R code below is an example on how to do that:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">paneljudge</span><span class="p">)</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">250523</span><span class="p">)</span><span class="w"> </span><span class="c1"># For reproducibility</span>

<span class="c1"># Read the data</span>
<span class="n">markers_info</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.delim</span><span class="p">(</span><span class="s">&quot;markers.tsv&quot;</span><span class="p">,</span><span class="n">stringsAsFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">F</span><span class="p">)[,</span><span class="m">-1</span><span class="p">]</span>
<span class="n">freq_values</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.delim</span><span class="p">(</span><span class="s">&quot;frequencies.tsv&quot;</span><span class="p">,</span><span class="n">stringsAsFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">F</span><span class="p">)</span>

<span class="c1"># Prepare the necessary data structures</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">markers_info</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">markers_info</span><span class="o">$</span><span class="n">Amplicon_name</span>
<span class="n">markers_info</span><span class="o">$</span><span class="n">pos</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">markers_info</span><span class="o">$</span><span class="n">pos</span><span class="p">)</span>
<span class="n">markers_info</span><span class="o">$</span><span class="n">chrom</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">markers_info</span><span class="o">$</span><span class="n">chrom</span><span class="p">)</span>

<span class="n">frequency_values</span><span class="o">&lt;-</span><span class="nf">list</span><span class="p">(</span><span class="s">&quot;ESEA&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">data.matrix</span><span class="p">(</span><span class="n">freq_values</span><span class="p">[,</span><span class="m">-1</span><span class="p">]))</span>
<span class="nf">dimnames</span><span class="p">(</span><span class="n">frequency_values</span><span class="o">$</span><span class="n">ESEA</span><span class="p">)[[</span><span class="m">1</span><span class="p">]]</span><span class="o">&lt;-</span><span class="n">freq_values</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span>
<span class="nf">dimnames</span><span class="p">(</span><span class="n">frequency_values</span><span class="o">$</span><span class="n">ESEA</span><span class="p">)[[</span><span class="m">2</span><span class="p">]]</span><span class="o">&lt;-</span><span class="nf">names</span><span class="p">(</span><span class="n">freq_values</span><span class="p">[,</span><span class="m">-1</span><span class="p">])</span>
<span class="nf">attr</span><span class="p">(</span><span class="n">frequency_values</span><span class="p">,</span><span class="s">&quot;split_type&quot;</span><span class="p">)</span><span class="o">&lt;-</span><span class="s">&quot;data.frame&quot;</span>
<span class="nf">attr</span><span class="p">(</span><span class="n">frequency_values</span><span class="p">,</span><span class="s">&quot;split_labels&quot;</span><span class="p">)</span><span class="o">&lt;-</span><span class="nf">data.frame</span><span class="p">(</span><span class="s">&quot;Country&quot;</span><span class="o">=</span><span class="s">&quot;ESEA&quot;</span><span class="p">,</span><span class="n">stringsAsFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">F</span><span class="p">)</span>

<span class="c1"># Save the data structures for future use</span>
<span class="nf">save</span><span class="p">(</span><span class="s">&quot;markers_info&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s">&quot;markers.RData&quot;</span><span class="p">)</span>
<span class="nf">save</span><span class="p">(</span><span class="s">&quot;frequency_values&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s">&quot;frequencies.RData&quot;</span><span class="p">)</span>

<span class="c1"># Simulate n genotype pairs, r in rs, and k in ks</span>
<span class="c1"># Relatedness values</span>
<span class="n">rs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;0.01&quot;</span><span class="o">=</span><span class="m">0.01</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;0.25&quot;</span><span class="o">=</span><span class="m">0.25</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;0.50&quot;</span><span class="o">=</span><span class="m">0.50</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;0.75&quot;</span><span class="o">=</span><span class="m">0.75</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;0.99&quot;</span><span class="o">=</span><span class="m">0.99</span><span class="p">)</span>
<span class="c1"># Switch rates</span>
<span class="n">ks</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;1&#39;</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;5&#39;</span><span class="o">=</span><span class="m">5</span><span class="p">,</span><span class="s">&#39;10&#39;</span><span class="o">=</span><span class="m">10</span><span class="p">,</span><span class="s">&#39;50&#39;</span><span class="o">=</span><span class="m">50</span><span class="p">)</span>

<span class="c1"># Number of pairs to per simulate per r in rs and k in ks</span>
<span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">100</span>

<span class="c1"># Run the simulations</span>
<span class="n">ds</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">markers_info</span><span class="o">$</span><span class="n">distances</span><span class="w"> </span><span class="c1"># Distances</span>
<span class="n">mles_CIs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lapply</span><span class="p">(</span><span class="n">frequency_values</span><span class="p">,</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">lapply</span><span class="p">(</span><span class="n">ks</span><span class="p">,</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">k</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nf">lapply</span><span class="p">(</span><span class="n">rs</span><span class="p">,</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="nf">lapply</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="n">n</span><span class="p">,</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">sim_Ys</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">simulate_Ys</span><span class="p">(</span><span class="n">fs</span><span class="p">,</span><span class="w"> </span><span class="n">ds</span><span class="p">,</span><span class="w"> </span><span class="n">k</span><span class="p">,</span><span class="w"> </span><span class="n">r</span><span class="p">)</span><span class="w"> </span><span class="c1"># Stimulate data</span>
<span class="w">        </span><span class="n">krhat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">estimate_r_and_k</span><span class="p">(</span><span class="n">fs</span><span class="p">,</span><span class="w"> </span><span class="n">ds</span><span class="p">,</span><span class="w"> </span><span class="n">Ys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sim_Ys</span><span class="p">)</span><span class="w"> </span><span class="c1"># Estimate k and r</span>
<span class="w">        </span><span class="n">CIs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_r_and_k_CIs</span><span class="p">(</span><span class="n">fs</span><span class="p">,</span><span class="w"> </span><span class="n">ds</span><span class="p">,</span><span class="w"> </span><span class="n">krhat</span><span class="p">[</span><span class="s">&#39;khat&#39;</span><span class="p">],</span><span class="w"> </span><span class="n">krhat</span><span class="p">[</span><span class="s">&#39;rhat&#39;</span><span class="p">])</span><span class="w"> </span><span class="c1"># CIs</span>
<span class="w">        </span><span class="nf">return</span><span class="p">(</span><span class="nf">cbind</span><span class="p">(</span><span class="n">krhat</span><span class="p">,</span><span class="w"> </span><span class="n">CIs</span><span class="p">))</span><span class="w"> </span><span class="c1"># Return mles and CIs</span>
<span class="w">      </span><span class="p">})</span>
<span class="w">    </span><span class="p">})</span>
<span class="w">  </span><span class="p">})</span>
<span class="p">})</span>

<span class="c1"># Save the results</span>
<span class="nf">save</span><span class="p">(</span><span class="n">mles_CIs</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;mles_CIs.RData&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="windows.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Discovery of potential microhaplotype markers in <em>Plasmodium vivax</em></p>
      </div>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#environment-and-data">Environment and data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-the-data">Loading the data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#load-the-genomic-data">Load the genomic data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-exploration">Data exploration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#candidate-selection">Candidate selection</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utility-functions">Utility functions</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#greedy">Greedy</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evenly-spaced">Evenly spaced</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#worked-example">Worked example</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Siegel, SV et al.
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>